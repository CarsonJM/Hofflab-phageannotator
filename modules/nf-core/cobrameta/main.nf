process COBRAMETA {
    tag "$meta.id"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/cobra-meta:1.2.3--10daf775a65e0312':
        'biocontainers/cobra-meta:1.2.3--pyhdfd78af_0' }"

    input:
    tuple val(meta) , path(fasta)
    tuple val(meta2), path(coverage)
    tuple val(meta3), path(query)
    tuple val(meta4), path(bam)
    val assembler
    val mink
    val maxk

    output:
    tuple val(meta), path("${prefix}_COBRA_extended.fasta.gz")      , emit: fasta
    tuple val(meta), path("${prefix}_COBRA_joining_summary.txt")    , emit: joining_summary
    tuple val(meta), path("${prefix}_log")                          , emit: log
    path "versions.yml"                                             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args    = task.ext.args ?: ''
    prefix      = task.ext.prefix ?: "${meta.id}"
    fasta_name  = fasta.getName().replace(".gz", "")
    """
    gunzip -c ${fasta} > ${fasta_name}
    tail ${virus_summary} -n +2 | awk '{print \$1}' > ${prefix}_viral_assemblies.txt
    tail ${coverage} -n +2 > ${prefix}_cobra_coverage.txt

    cobra-meta \\
        --fasta ${fasta_name} \\
        --coverage ${prefix}_cobra_coverage.txt \\
        --query ${prefix}_viral_assemblies.txt \\
        --mapping ${bam} \\
        --assembler ${assembler} \\
        --mink ${mink} \\
        --maxk ${maxk} \\
        --threads ${task.cpus} \\
        --output ${prefix} \\
        $args

    if [ -f ${prefix}/COBRA_extended.fasta ]; then
        mv ${prefix}/COBRA_extended.fasta ${prefix}_COBRA_extended.fasta
        gzip ${prefix}_COBRA_extended.fasta
        mv ${prefix}/COBRA_joining_summary.txt ${prefix}_COBRA_joining_summary.txt
        mv ${prefix}/log ${prefix}_log
    else 
        echo "" | gzip > ${prefix}_COBRA_extended.fasta.gz
        touch ${prefix}_COBRA_joining_summary.txt
        mv ${prefix}/log ${prefix}_log
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        cobra: \$(echo \$(cobra-meta --version 2>&1) | sed 's/^.*cobra v//' ))
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo "" | gzip > ${prefix}_COBRA_extended.fasta.gz
    touch ${prefix}_COBRA_joining_summary.txt
    touch ${prefix}_log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        cobra: \$(echo \$(cobra-meta --version 2>&1) | sed 's/^.*cobra v//' ))
    END_VERSIONS
    """
}
