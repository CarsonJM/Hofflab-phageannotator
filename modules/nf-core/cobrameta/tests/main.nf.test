nextflow_process {

    name "Test Process COBRAMETA"
    script "../main.nf"
    process "COBRAMETA"


    test("contigs.fasta.gz & coverage.tsv & viral_contigs & bam") {
        when {
            params {
                assembler   = 'metaspades'
                mink        = 27
                maxk        = 127
            }
            process {
                """
                input[0] = Channel.of (
                    [
                        [ id:'test', assembler:'metaspades' ],
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true)
                    ]
                )
                input[1] = Channel.of (
                    [
                        [ id:'test', assembler:'metaspades' ],
                        [
                            file("${projectDir}/assets/test-datasets/cobra/coverage.tsv", checkIfExists: true)
                        ]
                    ]
                )
                input[2] = Channel.of (
                    [
                        [ id:'test', assembler:'metaspades' ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/tsv/test1_contigs_ids.tsv', checkIfExists: true)
                        ]
                    ]
                )
                input[3] = Channel.of (
                    [
                        [ id:'test', assembler:'metaspades' ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/bam/test1_contigs.sorted.bam', checkIfExists: true)
                        ]
                    ]
                )
                input[4] = params.assembler
                input[5] = params.mink
                input[6] = params.maxk
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(
                    process.out.fasta,
                    process.out.joining_summary,
                    process.out.versions,
                    file(process.out.log[0][1]).name
                    ).match()
                }
            )
        }
    }


    test("contigs.fasta.gz & coverage.tsv & viral_contigs & bam - stub") {
        options "-stub"
        when {
            params {
                assembler   = 'metaspades'
                mink        = 27
                maxk        = 127
            }
            process {
                """
                input[0] = Channel.of (
                    [
                        [ id:'test', assembler:'metaspades' ],
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true)
                    ]
                )
                input[1] = Channel.of (
                    [
                        [ id:'test', assembler:'metaspades' ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/coverage/test1_contigs.coverage.stats.txt', checkIfExists: true)
                        ]
                    ]
                )
                input[2] = Channel.of (
                    [
                        [ id:'test', assembler:'metaspades' ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/tsv/test1_contigs_ids.tsv', checkIfExists: true)
                        ]
                    ]
                )
                input[3] = Channel.of (
                    [
                        [ id:'test', assembler:'metaspades' ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/bam/test1_contigs.sorted.bam', checkIfExists: true)
                        ]
                    ]
                )
                input[4] = params.assembler
                input[5] = params.mink
                input[6] = params.maxk
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
