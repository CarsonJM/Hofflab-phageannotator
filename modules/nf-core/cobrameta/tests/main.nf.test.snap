{
    "contigs.fasta.gz & coverage.tsv & viral_contigs & bam": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "assembler": "metaspades"
                    },
                    "test_COBRA_extended.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "assembler": "metaspades"
                    },
                    "test_COBRA_joining_summary.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                "versions.yml:md5,6830d6d3da149b4eebd6e66f2346ec92"
            ],
            "test_log"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-07-24T10:23:09.436497248"
    },
    "contigs.fasta.gz & coverage.tsv & viral_contigs & bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "assembler": "metaspades"
                        },
                        "test_COBRA_extended.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "assembler": "metaspades"
                        },
                        "test_COBRA_joining_summary.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "assembler": "metaspades"
                        },
                        "test_log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,6830d6d3da149b4eebd6e66f2346ec92"
                ],
                "fasta": [
                    [
                        {
                            "id": "test",
                            "assembler": "metaspades"
                        },
                        "test_COBRA_extended.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "joining_summary": [
                    [
                        {
                            "id": "test",
                            "assembler": "metaspades"
                        },
                        "test_COBRA_joining_summary.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "log": [
                    [
                        {
                            "id": "test",
                            "assembler": "metaspades"
                        },
                        "test_log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,6830d6d3da149b4eebd6e66f2346ec92"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-07-24T10:23:32.990113055"
    }
}